### PR TITLE
Detecting DC from the expression

### PR DESCRIPTION
**Pull request by studanshu**

### CODE REVIEW

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java [hunk 1]
```diff
@@ -47,16 +47,17 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: please use explicit imports

File: ArgusCore/src/main/java/com/salesforce/dva/argus/util/AlertUtils.java [hunk 2]
```diff
@@ -6,14 +6,15 @@
 import org.slf4j.LoggerFactory;
 
 import java.text.MessageFormat;
-import java.util.List;
+import java.util.*;
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/util/AlertUtils.java]: please use explicit import

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java [hunk 3]
```diff
@@ -227,5 +228,33 @@ private long _getDatapointsAcrossMetrics(List<Metric> metrics) {
 		}
 		return dataPointsSize;
 	}
+
+	public List<String> getDCFromExpression(String expression) {
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: where will this function be used ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java [hunk 4]
```diff
@@ -227,5 +228,33 @@ private long _getDatapointsAcrossMetrics(List<Metric> metrics) {
...
+		Set<String> DC = new HashSet<String>();
+		Matcher m;
+		ArrayList<String> patterns = new ArrayList<>(Arrays.asList( "\\.(?i)(CAP|CRZ|SFZ|XRD|CDG|CDU|CHI|CHX|CRD|DFW|FRA|FRF|HND|IA2|IAD|LON|ORD|PAR|PH2|PHX|PRD|SFM|SPZ|SYD|UKB|WAS|WAX|XRD|YHU|YUL)\\.",
+				":argus\\."));
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: This is open source repo, we should not expose our datacenter names here.  @dilipdevaraj-sfdc thoughts ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java (hunk 3 above)
Comment (studanshu) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: @nkunal to identify datalag per dc, we would need to know from which dc was the metric sent.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java [hunk 5]
```diff
@@ -227,5 +228,33 @@ private long _getDatapointsAcrossMetrics(List<Metric> metrics) {
...
+
+		if (expression == null || expression.isEmpty()) {
+			_logger.error("Expression either null or empty. Cannot retrive DC from the expression. Returning default value PRD.");
+			DC.add("PRD");
```
[the remark this answers is not in the document]
Comment (studanshu) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: @dilipdevaraj-sfdc could you please tell which config file should these parameter go into?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java (hunk 4 above)
Comment (studanshu) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: @nkunal @dilipdevaraj-sfdc made changes

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java [hunk 6]
```diff
@@ -227,5 +236,35 @@ private long _getDatapointsAcrossMetrics(List<Metric> metrics) {
...
+				m = Pattern.compile(patterns.get(0)).matcher(currentQuery.getScope());
+				while (m.find()) DC.add(m.group().substring(1, m.group().length() - 1).toUpperCase());
+
+				// If it matches ":argus., then it's argus metrics emitted by PRD."
+				m = Pattern.compile(patterns.get(1)).matcher(currentQuery.getScope());
+				if(m.find()) DC.add(defaultDC);
+			}
+		}
+		if (DC.size() == 0) {
+			_logger.info("Unable to identify DC from expression: " + expression +" . Returning default value PRD.");
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/metric/DefaultMetricService.java]: remove PRD from log

File: ArgusCore/src/main/java/com/salesforce/dva/argus/system/SystemConfiguration.java [hunk 7]
```diff
@@ -1,297 +1,300 @@
...
+            return _defaultValue;
+        }
+
+        private String key() {
+            return _name;
+        }
+    }
+
+}
+/* Copyright (c) 2016, Salesforce.com, Inc.  All rights reserved. */
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/system/SystemConfiguration.java]: Why is this whole file showing up as diff ?

**Pull request merged**